docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+| ------------- | ------------- | ------------- |
+| `getProps()` | `table` | returns a type representation of tables (e.g. {name = "John"} will return {[string] = "string"}) |
+| `getIndexer()` | `table` | returns a type representation of arrays (e.g. {1, "hi", 3} will return {[number] = "number" \| "string"}) |
+
+#### Class
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getName()` | `table` | returns the name of self's class |
+| `getParent()` | `lType` | returns lType userdata of self's parent |
+
+</details>
+
+### Implementation
+
+The implementation of user-defined type functions can be broken down into the following parts:
+
+**AST**: Introduce a new AST node for type functions called `AstStatTypeFunction`.
+
+**Parser**: When parsing TypeAliases, if the next lexeme is "function", parse it into the `AstStatTypeFunction` node.
+
+**Constraint Generator**: Generate a `TypeAliasExpansionConstraint` when visiting the `AstStatTypeFunction` node.
+
+**Constraint Solver**: Generate a new `ReduceConstraint` and append it to the list of unsolved constraints when reducing `TypeAliasExpansionConstraint`.
+
+**Reduction Constraint**: Serialize the function arguments into `lType` and execute the body of the function. Reduce to the deserialized version of the function return value.
+
+**lType**: Using the Lua API, create a library that interfaces between C++ and Luau and supports all of the function calls, serialization, and deserialization.
+
+## Drawback
+
+Type functions are handled at the analysis time, while `lType` is an implementation in the runtime. As a result, the proposed design causes the analysis time to be dependent on Luau's runtime to reduce user-defined type functions. This is generally discouraged as it is best to isolate the compile time, analysis time, and runtime from each other for the purpose of maintaining a clean separation of concerns, which helps minimize side effects and dependencies across different phases of the program execution and improves the modularity of the compiler. Overlaps between the analysis time and runtime can lead to code that is more complex and harder to manage, potentially increasing the risk of bugs and making the outcomes less predictable.
+
+The build / analysis times will also be negatively impacted by this feature as reducing type functions takes variable amount of time based on the program. The larger the type function, the longer it will take to reduce it in the constraint solver.
+
+## Alternatives
+
+### `table` Runtime Representation
+
+Currently, the runtime representation of types is a userdata called `lType`. Another representation is to use the already-existing type in Luau `table`; instead of serializing types into `lType`, we can serialize them into tables with predefined properties. For instance, the representation for a string singleton `"abc"` could be `{type = "stringSingleton", value = "abc"}`. So instead of writing:
```

Comment:
I would much rather have this runtime representation!

Reply:
I'm curious as to why. I originally wrote it that way in the earliest prototype for type functions, but I directed Jeremy to implement it as a userdata because we are, fundamentally, embedding types from the type system into these VM instances and that's... really what userdata is _for_. I do think that the corresponding C++, if we go the userdata route, should probably be a little OO hierarchy for the types, which isn't quite what's described here. But I envisioned something akin to the instance hierarchy we embed from Roblox.

One question here is which of these would be easier to add typing for (i.e. actually type checking the type functions), and I think the userdata approach might be somewhat easier? But at the end of the day, they're probably both going to lean pretty heavily on refinements.

Reply:
mmm. I guess it can go either way, though since I suggested using `==`/`>=`/`<=` for subtyping, the only way that'd realistically work is by using userdata. Let's go with that then.